discovery/linode/linode.go

```diff
@@ -0,0 +1,261 @@
+	return NewDiscovery(c, opts.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+// Discovery periodically performs Linode requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client       *linodego.Client
+	port         int
+	tagSeparator string
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port:         conf.Port,
+		tagSeparator: conf.TagSeparator,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "linode_sd", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	client := linodego.NewClient(
+		&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		},
+	)
+	client.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version))
+	d.client = &client
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"linode",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "Linode",
+	}
+
+	// Gather all linode instances.
+	instances, err := d.client.ListInstances(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Gather detailed IP address info for all IPs on all linode instances.
+	detailedIPs, err := d.client.ListIPAddresses(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range instances {
+		if len(instance.IPv4) == 0 {
+			continue
+		}
+
+		var (
+			privateIPv4, publicIPv4, publicIPv6             string
+			privateIPv4RDNS, publicIPv4RDNS, publicIPv6RDNS string
+			backupsStatus                                   string
+			extraIPs                                        []string
+		)
+
+		for _, ip := range instance.IPv4 {
+			for _, detailedIP := range detailedIPs {
+				if detailedIP.Address != ip.String() {
+					continue
+				}
+
+				if detailedIP.Public && publicIPv4 == "" {
+					publicIPv4 = detailedIP.Address
+
+					if detailedIP.RDNS != "" && detailedIP.RDNS != "null" {
+						publicIPv4RDNS = detailedIP.RDNS
+					}
+				} else if !detailedIP.Public && privateIPv4 == "" {
+					privateIPv4 = detailedIP.Address
+
+					if detailedIP.RDNS != "" && detailedIP.RDNS != "null" {
+						privateIPv4RDNS = detailedIP.RDNS
+					}
+				} else {
+					extraIPs = append(extraIPs, detailedIP.Address)
+				}
+			}
+		}
+
+		if instance.IPv6 != "" {
```

Comment:
There is no logic to handle the extra ipv6?

Reply:
The IPv6 in the instance endpoint response is the instance's default SLAAC address. Instances can have multiple IP addresses, and Linode offers 2 kinds of additional IPv6:
https://www.linode.com/docs/guides/an-overview-of-ipv6-on-linode/#additional-ipv6-addresses

IPv6 pools are accessible from every Linode within the datacenter, and thus don't really "belong" to a Linode. IPs from the pool can be assigned to instances at will, but only the pool is tracked in the API, and we don't have an API driven way to know about extra IPs from such a pool being assigned to a Linode, so I don't think this is possible. 

IPv6 routed ranges are _routed_ to individual instances, but traffic for any IPs within the routed range are forwarded to the instance's `/128` SLAAC address anyway, so similarly, we currently have no API driven way to know about extra IPs assigned to an instance. 

There are endpoints in the API to list all IPv6 pools/ranges, but we have no way to tie an IPv6 address from those ranges to an individual instance. =(